Fix empty cctor

RikkiGibson opened the pull request:
Closes #42985 

~~Solves the issue by excluding the .cctor from the public members list and waiting until emit to decide whether to synthesize the constructor.~~

Solves the issue without changing the conditions where we include the synthesized static constructor in GetMembers(). Instead we decide selectively to not emit the synthesized static constructor if the field initializers are only assigning the default values.

- Some room for improvement in sharing logic to decide to skip emitting a method (similar to the default struct constructor or a partial method with no implementation part).
- ~~We could try to skip emitting even an explicitly declared static constructor if we find that the lowered implementation is empty. I feel that this is a bit more risky/difficult to get right.~~ We will only eliminate synthesized static constructors.

RikkiGibson commented:
This is still not ready to merge (some sloppy copy-pasting and incomplete checks around 'extern static' constructors). But I would like to open it for input. I am unsure whether to skip the static constructor based on the lowered bound tree, or whether to skip it based on the emitted body containing only `IlOpCode.Ret`.

- Using the lowered bound tree has a problem where we have to do some guesswork to decide if the method body would result in an "empty" IL body. This includes stuff like filtering out sequence point nodes and filtering out unreachable nodes, i.e. `if (false) DoSomething();`.

- Using the emitted body from `GenerateMethodBody()` has a problem of unwanted side effects such as attaching unwanted test data or pdb entries to the compilation.

/cc @AlekseyTs.

AlekseyTs commented:
We should never eliminate an explicitly declared  constructor. Also, for the first phase we can simply handle the scenario where bound initializers (after initial binding) produce obviously "default" values, i.e. no conditional logic, etc.

RikkiGibson commented:
I misread https://github.com/dotnet/roslyn/issues/42985#issuecomment-612210106 to mean that we could even eliminate explicitly declared `.cctor`s as long as the implementation is empty. Eliminating only a synthesized `.cctor` by checking the bound initializers is not too hard. I had it working that way in one of the commits along the way in this PR. Will update the PR accordingly.

RikkiGibson commented:
The first thing that comes to mind for me is a user may expect to be able to add an empty static constructor and set a breakpoint in it just to debug something about their program’s initialization.

Eliminating the explicitly declared but empty static constructors is something we could conceivably do, but it is riskier and less compelling. We probably wouldn’t be making this change for the synthesized static constructors if not for the `string field = null!` idiom.

jaredpar commented:
> The first thing that comes to mind for me is a user may expect to be able to add an empty static constructor and set a breakpoint in it just to debug something about their program’s initialization

That makes sense. If a user explicitly adds one then we should not be deleting it. I was thinking more of the cases where we implicitly generate one for them.

cston commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1102,12 +1102,62 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
+                            if (value is BoundDefaultExpression)
+                            {
+                                continue;
+                            }
```
If the type is a value type, shouldn't we emit the .cctor?
```
private static S _s = default;
```

RikkiGibson replied:
We expect the runtime to zero out such fields. The optimization that removes redundant initializers will remove this initializer. See the empty .cctor in this example: https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHAZgAJYEBXAYxkoGUDKBvUy7yiytAHZMAZmjgYAJgG5SAX1Kk++SgGEOXHkoIA2FmyhsAvJQlxhEWhhgySsoA==

cston commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1102,12 +1102,62 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
+                    if (methodSymbol is SynthesizedStaticConstructor)
+                    {
+                        foreach (var initializer in initializers)
+                        {
```
>{ [](start = 24, length = 1)

Consider extracting the loop body to a helper method and use `return initializers.Any(i => HasStatements(i));`.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/SynthesizedStaticConstructorTests.cs:
```diff
@@ -265,6 +265,24 @@ class C
+            // Although we do not emit the synthesized static constructor, the source type symbol will still appear to have one
+            Assert.True(HasSynthesizedStaticConstructor(typeSymbol));
+            Assert.True(IsBeforeFieldInit(typeSymbol));
+        }
```
>} [](start = 8, length = 1)

Please test:
```
static S s1 = default;
static object s2 = default(S);
```

jaredpar replied:
Consider using a `Theory` to keep the test compact.

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/NamedTypeSymbolAdapter.cs:
```diff
@@ -668,6 +668,11 @@ IEnumerable<Cci.IMethodDefinition> Cci.ITypeDefinition.GetMethods(EmitContext co
             foreach (var method in this.GetMethodsToEmit())
             {
                 Debug.Assert((object)method != null);
+                if (method is SynthesizedStaticConstructor && context.Module.GetMethodBody(method) is null)
```
>is SynthesizedStaticConstructor [](start = 27, length = 31)

I searched for ```is SynthesizedStaticConstructor``` within the project and couldn't find any match. However, it is unlikely that we don't already do an equivalent check somewhere. I think that kind of check would be preferred over a type check.

AlekseyTs replied:
See for example MethodSymbol.IsImplicitConstructor or MethodSimbol.IsImplicitInstanceConstructor. I suggest adding similar helper next to them and use it.

---
In reply to: [410527156](https://github.com/dotnet/roslyn/pull/43234#discussion_r410527156) [](ancestors = 410527156)

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/NamedTypeSymbolAdapter.cs:
```diff
@@ -668,6 +668,11 @@ IEnumerable<Cci.IMethodDefinition> Cci.ITypeDefinition.GetMethods(EmitContext co
             foreach (var method in this.GetMethodsToEmit())
```
>GetMethodsToEmit [](start = 40, length = 16)

I think that the filtering should be done by ```GetMethodsToEmit``` method. The filtering performed by this method is full vs. reference assembly. Other consumers of ```GetMethodsToEmit``` (existing or future) should not see the filtered out constructor.

RikkiGibson replied:
Not yet sure how to implement this part. I am checking if the synthesized static constructor will be emitted using PEModuleBuilder,GetMethodBody. Some of the call sites don't have a PEModuleBuilder around, so some other way to check the constructor is probably necessary.

Maybe I need to add a property on the synthesized constructor symbol which forces binding the static initializers, if it hasn't happened yet?

RikkiGibson replied:
Because the other place where we make this check enumerates all members, not just methods, I ended up doing this by factoring out a ShouldEmit helper for methods.

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:
```diff
@@ -263,6 +263,7 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
                                         //       but adds synthesized instance constructors, Roslyn adds both
                                         var method = (MethodSymbol)member;
                                         if (method.IsDefaultValueTypeConstructor() ||
+                                            (method is SynthesizedStaticConstructor && GetMethodBody(method) is null) ||
```
>(method is SynthesizedStaticConstructor && GetMethodBody(method) is null) [](start = 44, length = 73)

Please add a comment explaining what is happening

AlekseyTs replied:
>method is SynthesizedStaticConstructor  [](start = 45, length = 39)

Same comment about the type check

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1102,12 +1102,62 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
+                // "meaningful" statements. If it only initializes fields to default values we will skip emitting it.
+                bool isEmptySynthesizedStaticConstructor(ImmutableArray<BoundInitializer> initializers)
+                {
+                    if (methodSymbol is SynthesizedStaticConstructor)
```
>methodSymbol is SynthesizedStaticConstructor [](start = 24, length = 44)

Same comment about the type check

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1102,12 +1102,62 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                 // Don't lower if we're not emitting or if there were errors. 
                 // Methods that had binding errors are considered too broken to be lowered reliably.
-                if (_moduleBeingBuiltOpt == null || hasErrors)
+                if (_moduleBeingBuiltOpt == null || hasErrors || isEmptySynthesizedStaticConstructor(processedInitializers.BoundInitializers))
```
>isEmptySynthesizedStaticConstructor(processedInitializers.BoundInitializers) [](start = 65, length = 76)

It feels like the check is making us to return too early. I think we shouldn't emit the body, but should still go through lowering and other analysis we might do.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConstructorInitTests.cs:
```diff
@@ -607,5 +609,183 @@ public void BadExpressionConstructor()
+
+class C
+{
+    static S field1 = default;
```
>default [](start = 22, length = 7)

Please also test:
- ```new S()``` and similar scenarios for nullable value type.
- Initializing a nullable value type with ```null```

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:
```diff
@@ -557,6 +557,7 @@ internal SomeAttribute()
 
     static SomeAttribute()
     {
+        Console.WriteLine();
```
What is the purpose of this change?

RikkiGibson replied:
artifact of the approach which excluded empty user-defined static constructors from emit. reverting.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2599,7 +2599,7 @@ class C
     int x = 1;
 
-    static C() { }
+    static C() { Console.WriteLine(); }
```
>Console.WriteLine(); [](start = 17, length = 20)

What is the purpose of this change? And below as well.

RikkiGibson replied:
artifact of the approach which excluded empty user-defined static constructors from emit. reverting.

AlekseyTs commented:
>     public static int Count = 0;

Is the new baseline exactly the same as for the case when this field doesn't have an initializer?

---
Refers to: src/Compilers/CSharp/Test/Emit/PDB/PDBAsyncTests.cs:37 in bd74323. [](commit_id = bd7432355cbe0af03e82186d65ac3a31615cd73a, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/SynthesizedStaticConstructorTests.cs:
```diff
@@ -265,6 +265,24 @@ class C
+
+            // Although we do not emit the synthesized static constructor, the source type symbol will still appear to have one
+            Assert.True(HasSynthesizedStaticConstructor(typeSymbol));
+            Assert.True(IsBeforeFieldInit(typeSymbol));
```
>IsBeforeFieldInit [](start = 24, length = 17)

Shouldn't this be false since we won't be emitting this constructor. We should have metadata equivalent to a situation when we didn't even synthesize the constructor symbol in the first place.

RikkiGibson replied:
I decided to keep existing behavior around when a synthesized static constructor symbol is included, because removing it  requires some churn around flow analysis of static initializers, etc. and it could be a breaking behavior change for outside consumers.

In addition, it seemed impractical to decide to include the synthesized static constructor symbol based on whether there are "meaningful" static initializers. That requires binding the initializers, which we want to avoid doing just for a call to TypeSymbol.GetMembers().

RikkiGibson replied:
I added a test which demonstrates that identical type IL is generated with or without the `null` initializer.

AlekseyTs replied:
> I added a test which demonstrates that identical type IL is generated with or without the null initializer.

Where is the test and what is its name?

---
In reply to: [411767188](https://github.com/dotnet/roslyn/pull/43234#discussion_r411767188) [](ancestors = 411767188)

RikkiGibson replied:
The test `StaticNullInitializerHasNoEffectOnTypeIL` in CodeGenConstructorInitTests.cs

AlekseyTs commented:
It looks like there are two code paths where we create an instance of SynthesizedStaticConstructor outside of BuildMembersAndInitializers. Please make sure we have tests to verify expected behavior for those code paths too.

AlekseyTs commented:
Done with review pass (iteration 9)

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:
```diff
@@ -263,6 +263,8 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
                                         var method = (MethodSymbol)member;
                                         if (method.IsDefaultValueTypeConstructor() ||
+                                            // Synthesized static constructors that will not be emitted should not be included in the SymbolToLocationMap
+                                            (method.IsImplicitStaticConstructor && GetMethodBody(method) is null) ||
```
>(method.IsImplicitStaticConstructor && GetMethodBody(method) is null) || [](start = 44, length = 72)

I am investigating a separate issue and this loop is causing similar issue, i.e. it is duplicates what Get...ToEmit APIs do and adjustments to them causing breaks here unless the logic is duplicated. I think we should rewrite it in terms of those APIs.

RikkiGibson commented:
Pushing to get a test run going. Haven't tried to fix GetMethodsToEmit behavior yet to reduce duplication of logic.

Also, instead of trying to handle every possible initializer that ends up lowering into a no-op, I'm just trying to handle the ones that really matter, particularly the scenario called out in the linked issue.

RikkiGibson commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1234,28 +1236,31 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                         }
 
-                        CSharpSyntaxNode syntax = methodSymbol.GetNonNullSyntaxNode();
+                        if (!(methodSymbol is SynthesizedStaticConstructor { ShouldEmit: false }))
```
This check is really a bit redundant, but it makes the VerifyMemberInIL calls behave as expected. Maybe we could ensure that if a method is not in "GetMethodsToEmit" we do not generate a body for it. Or we could require VerifyMemberInIL to check both the module builder and GetMethodsToEmit to decide if the "member is in IL" or not.

AlekseyTs replied:
> This check is really a bit redundant, but it makes the VerifyMemberInIL calls behave as expected. Maybe we could ensure that if a method is not in "GetMethodsToEmit" we do not generate a body for it. Or we could require VerifyMemberInIL to check both the module builder and GetMethodsToEmit to decide if the "member is in IL" or not.

I think that preferred way to assert that a method definition is not emitted is to use symbols. CompileAndVerify has symbolValidator exactly for that purpose. A method definition can be emitted even when it has no body, we actually need to verify that we didn't even emit a definition. Given that, I am not really concerned if VerifyMemberInIL misbehaves for this scenario, it is just a test helper and doesn't have to handle all possible scenarios. However, it does make sense to not attempt to emit the body if we are not planning to emit a declaration.

---
In reply to: [412583567](https://github.com/dotnet/roslyn/pull/43234#discussion_r412583567) [](ancestors = 412583567)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedStaticConstructor.cs:
```diff
@@ -371,5 +372,69 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
+
+#nullable enable
+        private readonly object _syncRoot = new object();
+        private Binder.ProcessedFieldInitializers _processedInitializers;
```
>Binder.ProcessedFieldInitializers _processedInitializers; [](start = 16, length = 57)

It is a bad idea to cache this on the symbol. Perhaps, I wasn't clear during our offline conversation. I  was talking about caching a bool representing whether to emit the method or not. I don't think we should be concerned much about reusing result of binding. The reason why caching Binder.ProcessedFieldInitializers is a bad idea is that it holds on to bound nodes, those refer to syntax nodes, making it impossible to unload syntax trees. IDE will be very unhappy. Symbols should never keep syntax nodes alive, unless the symbols are very short lived.

RikkiGibson replied:
Was the thought for it to work something like this?

```cs
private bool? _shouldEmit;
internal bool ShouldEmit(ImmutableArray<BoundInitializer> boundInitializersOpt = default)
{
    if (_shouldEmit != null)
        return _shouldEmit.GetValueOrDefault();

    if (boundInitializersOpt.IsDefault)
        boundInitializersOpt = BindFieldInitializers(...);

    _shouldEmit = CheckFieldInitializers(boundInitializersOpt, ...);
    return _shouldEmit.GetValueOrDefault();
}
```

AlekseyTs replied:
> Was the thought for it to work something like this?

Yes, but bool? is not thread-safe, we usually using TreeState enim or something like that, we have a type and examples


---
In reply to: [413184581](https://github.com/dotnet/roslyn/pull/43234#discussion_r413184581) [](ancestors = 413184581)

RikkiGibson commented:
I found that the only time in the emit tests where we actually have to bind the initializers with `ShouldEmit()` is for EnC.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -449,6 +449,13 @@ internal SynthesizedEntryPointSymbol GetScriptEntryPoint()
         }
 
+#nullable enable
+        internal SynthesizedStaticConstructor? GetSynthesizedStaticConstructor()
```
>GetSynthesizedStaticConstructor [](start = 47, length = 31)

It doesn't look like this API is used anywhere. Please remove it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedStaticConstructor.cs:
```diff
@@ -371,5 +371,56 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
         }
+
+#nullable enable
+        private ThreeState _shouldEmit = ThreeState.Unknown;
```
>private ThreeState _shouldEmit = ThreeState.Unknown; [](start = 8, length = 52)

Please place this field declaration at the top of the class. We usually place them before a constructor. This makes it easier to see in a quick glance what data we already have, etc.

AlekseyTs replied:
>_shouldEmit [](start = 27, length = 11)

We usually use "lazy" prefix to indicate that the value is calculated lazily.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedStaticConstructor.cs:
```diff
@@ -371,5 +371,56 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
+            return shouldEmit;
+        }
+
+        private bool InitializeShouldEmit(ImmutableArray<BoundInitializer> boundInitializersOpt = default)
```
>InitializeShouldEmit [](start = 21, length = 20)

This function doesn't initialize, it calculates. Consider renaming

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedStaticConstructor.cs:
```diff
@@ -371,5 +371,56 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
+                    return true;
+                }
+
+                if (!value.IsDefaultValue())
```
>if (!value.IsDefaultValue()) [](start = 16, length = 28)

Stopping at handling just these simple cases is an option and we certainly don't have to expand it in this PR. This is more a design question and I think it is worth having an offline discussion at some point.

AlekseyTs commented:
> It looks like there are two code paths where we create an instance of SynthesizedStaticConstructor outside of BuildMembersAndInitializers. Please make sure we have tests to verify expected behavior for those code paths too.

Was this addressed?

---
In reply to: [615538957](https://github.com/dotnet/roslyn/pull/43234#issuecomment-615538957) [](ancestors = 615538957)

AlekseyTs commented:
>     public static int Count = 0;

> Is the new baseline exactly the same as for the case when this field doesn't have an initializer?

There was no response to this question.

---
In reply to: [615535416](https://github.com/dotnet/roslyn/pull/43234#issuecomment-615535416) [](ancestors = 615535416)

---
Refers to: src/Compilers/CSharp/Test/Emit/PDB/PDBAsyncTests.cs:37 in bd74323. [](commit_id = bd7432355cbe0af03e82186d65ac3a31615cd73a, deletion_comment = False)

AlekseyTs commented:
Done with review pass (iteration 15)

RikkiGibson commented:
> Is the new baseline exactly the same as for the case when this field doesn't have an initializer?

Modified `TestAsyncDebug` to demonstrate that this is the case.

> It looks like there are two code paths where we create an instance of SynthesizedStaticConstructor outside of BuildMembersAndInitializers. Please make sure we have tests to verify expected behavior for those code paths too.

Added some tests specifically to verify the .cctor that gets created for a const decimal field and for a non-capturing lambda.

RikkiGibson commented:
It also sounds like I should also start using symbolValidator to check whether or not a .cctor is present. Will make a pass to fix that when I get the chance.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConstructorInitTests.cs:
```diff
@@ -592,6 +594,58 @@ .maxstack  4
+            var source = @"
+class C
+{
+    const decimal d1 = 0.1m;
```
>0.1m [](start = 23, length = 4)

Let's also add a test with ```0m```. Will it be treated as default and eliminate the constructor?

RikkiGibson replied:
I found that `0m` was treated differently than `0` or `default`. That seems like a bug.

RikkiGibson replied:
(actually, it seems like the precision of `0.0m` made it a non-default value. `0m` is treated as expected.)

AlekseyTs commented:
Done with review pass (iteration 17)

RikkiGibson commented:
Could I please get a second review @dotnet/roslyn-compiler

cston commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedStaticConstructor.cs:
```diff
@@ -371,5 +374,54 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
+                if (!(ContainingType is SourceMemberContainerTypeSymbol sourceType))
+                {
+                    return true;
+                }
```
Is this case for synthesized classes? Are we emitting a static constructor for those cases always?

RikkiGibson replied:
Based on the usages of SynthesizedStaticConstructor I think the only synthesized class to be concerned about is SynthesizedClosureEnvironment which has static fields for the display class and sometimes for the delegate.

cston replied:
Consider asserting `ContainingType` is `SynthesizedClosureEnvironment` so we catch other cases as synthesized class types are added.

---
In reply to: [415115832](https://github.com/dotnet/roslyn/pull/43234#discussion_r415115832) [](ancestors = 415115832)

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConstructorInitTests.cs:
```diff
@@ -607,5 +737,488 @@ public void BadExpressionConstructor()
+
+        [WorkItem(42985, "https://github.com/dotnet/roslyn/issues/42985")]
+        [Fact]
+        public void SkipSynthesizedStaticConstructor_07()
```
>SkipSynthesizedStaticConstructor_07 [](start = 20, length = 35)

Already covered by `_01`.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConstructorInitTests.cs:
```diff
@@ -607,5 +737,488 @@ public void BadExpressionConstructor()
+    static S? s3 = null;
+    static S? s4 = new S?();
+}";
+            // note: we could make the synthesized constructor smarter and realize that
```
>note: we could make the synthesized constructor smarter [](start = 15, length = 55)

Which of the above cases causes the .cctor to be emitted? (Based on the previous test, it looks like `= null` does not result in a .cctor.) Consider splitting the test into two, with just the case(s) that result in .cctor in one and the rest in the other.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConstructorInitTests.cs:
```diff
@@ -607,5 +737,488 @@ public void BadExpressionConstructor()
+{
+    static int i = 0;
+    static bool b = false;
+}";
```
Consider including all primitive types:
```
static sbyte s1 = 0;
static byte s2 = 0;
static short s3 = 0;
static ushort s4 = 0;
static int s5 = 0;
static uint s6 = 0;
static long s7 = 0;
static ulong s8 = 0;
static nint s9 = 0;
static nuint s10 = 0;
static bool s11 = false;
static char s12 = '\0';
static float s13 = 0f;
static double s14 = 0d;
static System.IntPtr s15 = (System.IntPtr)0;
static System.UIntPtr s16 = (System.UIntPtr)0;
static void* s17 = (void*)0;
```
And consider tests of the same types with `default` and `default(T)`.

RikkiGibson replied:
It seems the explicit pointer conversions require things to be emitted today, so i will factor those into their own test, but otherwise I am going to handle this by expanding `CodeGenTests.Regress530049`.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConstructorInitTests.cs:
```diff
@@ -607,5 +737,497 @@ public void BadExpressionConstructor()
+class C
+{
+    static S? s1 = default;
+    static S? s2 = new S?();
```
>static S? s2 = new S?(); [](start = 4, length = 24)

Strictly speaking we should split these two cases as well in order to verify that each of the scenarios results in an emitted constructor.
